include/execution.hpp

```diff
@@ -1922,17 +1918,60 @@ _PRAGMA_POP()
                 __types<exception_ptr>,
                 error_types_of_t<__t<_SenderId>, __types>>>>;
 
-      template <class _Sender2>
-        using __sends_done =
-          bool_constant<sender_traits<_Sender2>::sends_done>;
-
       static constexpr bool sends_done =
         __result_senders_t<
           __transform<
-            __q1<__sends_done>,
+            __q1<__let::__impl::__sends_done>,
             __right_fold<false_type, __q2<__let::__impl::__or>>>>::value;
     };
 
+  /////////////////////////////////////////////////////////////////////////////
+  // [execution.senders.adaptors.done_as_optional]
+  // [execution.senders.adaptors.done_as_error]
+  inline namespace __done_as_xxx {
+    inline constexpr struct __done_as_optional_t {
+      template <__single_typed_sender _Sender>
+        auto operator()(_Sender&& __sndr) const {
+          using __value_t = decay_t<__single_sender_value_t<_Sender>>;
+          return (_Sender&&) __sndr
+            | then([]<class _T>(_T&& __t) {
+                return optional<__value_t>{(_T&&) __t};
+              })
+            | let_done([]() noexcept {
+                return just(optional<__value_t>{});
+              });
+        }
+      __binder_back<__done_as_optional_t> operator()() const noexcept {
+        return {};
+      }
+    } done_as_optional {};
+
+    inline constexpr struct __done_as_error_t {
+      template <sender _Sender, __movable_value _Error>
+        auto operator()(_Sender&& __sndr, _Error __err) const {
+          return (_Sender&&) __sndr
+            | let_done([__err2 = (_Error&&) __err] () mutable {
+                return just_error((_Error&&) __err2);
+              });
+        }
+      template <__movable_value _Error>
+        auto operator()(_Error __err) const
+          -> __binder_back<__done_as_error_t, _Error> {
+          return {{}, {}, {(_Error&&) __err}};
+        }
+    } __done_as_error {};
+
+    template <__movable_value _Error, sender _Sender>
+      auto done_as_error(_Sender&& __sndr, _Error __err = _Error{}) {
+        return __done_as_error((_Sender&&) __sndr, (_Error&&) __err);
+      }
+
+    template <__movable_value _Error>
+      auto done_as_error(_Error __err) noexcept {
+        return __done_as_error((_Error&&) __err);
+      }
+  }
```

Comment:
why is done_as_error() an overload set while done_as_optional is a function object? 
I would have expected both to be function objects.

Reply:
It support the syntax: `done_as_error<Error>(sender)`, which default-constructs the error. Not terribly compelling, is it?

Reply:
No, I do not find that compelling.